src/uuid.cr

```diff
@@ -0,0 +1,217 @@
+# Universally Unique IDentifier.
+# Supports [RFC4122](https://www.ietf.org/rfc/rfc4122.txt) UUIDs and custom
+struct UUID
+  enum Variant # variants with 16 bytes.
+    Unknown    # Unknown (ie. custom, your own).
+    NCS        # Reserved by the NCS for backward compatibility.
+    RFC4122    # Reserved for RFC4122 Specification (default).
+    Microsoft  # Reserved by Microsoft for backward compatibility.
+    Future     # Reserved for future expansion.
+  end
+
+  enum Version  # RFC4122 UUID versions.
+    Unknown = 0 # Unknown version.
+    V1      = 1 # date-time and MAC address.
+    V2      = 2 # DCE security.
+    V3      = 3 # MD5 hash and namespace.
+    V4      = 4 # random.
+    V5      = 5 # SHA1 hash and namespace.
+  end
+
+  # Internal representation.
+  @bytes : StaticArray(UInt8, 16)
+
+  # Generates RFC 4122 v4 UUID.
+  def self.random(variant = Variant::RFC4122, version = Version::V4)
+    new_bytes = uninitialized UInt8[16]
+    Random.new.random_bytes(new_bytes.to_slice)
+
+    new(new_bytes, variant, version)
+  end
+
+  def self.empty
+    new(StaticArray(UInt8, 16).new(0_u8), UUID::Variant::NCS, UUID::Version::V4)
+  end
+
+  # Creates UUID from 16-bytes slice.
+  def initialize(slice : Slice(UInt8), variant = Variant::RFC4122, version = Version::V4)
+    raise ArgumentError.new "Invalid bytes length #{@bytes.size}, expected 16" unless slice.size == 16
+
+    @bytes = uninitialized UInt8[16]
+    slice.copy_to(@bytes.to_slice)
+
+    self.variant = variant
+    self.version = version
+  end
+
+  # Generates UUID from static 16-`bytes`, applying *version* and *variant* to the UUID.
+  def initialize(@bytes : StaticArray(UInt8, 16), variant : UUID::Variant = Variant::RFC4122, version : UUID::Version = Version::V4)
+    self.variant = variant
+    self.version = version
+  end
+
+  # Creates UUID from string `value`. See `UUID#decode(value : String)` for details on supported string formats.
+  def initialize(value : String, variant = Variant::RFC4122, version = Version::V4)
+    @bytes = uninitialized UInt8[16]
+    decode value
+
+    self.variant = variant
+    self.version = version
```

Comment:
We should have an option *not* to overwrite the variant and version of the string uuid.

Reply:
This is interesting I think you have a good point.
What happens if someone sends in a variant and versions and also options?

I am not sure this follows the general conventions of Crystal but what about making the parameters have a nil or another flag.

something like `new(string, variant: :no_overwrite)`

Reply:
No, by default the `variant` and `version` should be `nil` and then you use `self.variant = variant if variant`. But this shouldn't be a mutable struct.

We should have a helper method which takes a `StaticArray` and overwrites variant and version like I say, then use this before setting `@bytes` instead of having `variant=` and `version=`.